src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -559,33 +558,61 @@ private long writeRawLogTo(AnnotatedLargeText<?> text, long start, OutputStream
     }
 
     @GuardedBy("completed")
-    private transient LoadingCache<FlowNode,Optional<String>> logPrefixCache;
-    private @CheckForNull String getLogPrefix(FlowNode node) {
-        // TODO could also use FlowScanningUtils.fetchEnclosingBlocks(node).filter(FlowScanningUtils.hasActionPredicate(ThreadNameAction.class)),
-        // but this would not let us cache intermediate results
+    private transient Cache<FlowNode,Optional<String>> branchNameCache;  // TODO Consider making this a top-level FlowNode API
+
+    private Cache<FlowNode, Optional<String>> getBranchNameCache() {
         synchronized (completed) {
-            if (logPrefixCache == null) {
-                logPrefixCache = CacheBuilder.newBuilder().weakKeys().build(new CacheLoader<FlowNode,Optional<String>>() {
-                    @Override public @Nonnull Optional<String> load(FlowNode node) {
-                        if (node instanceof BlockEndNode) {
-                            return Optional.fromNullable(getLogPrefix(((BlockEndNode) node).getStartNode()));
-                        }
-                        ThreadNameAction threadNameAction = node.getAction(ThreadNameAction.class);
-                        if (threadNameAction != null) {
-                            return Optional.of(threadNameAction.getThreadName());
-                        }
-                        for (FlowNode parent : node.getParents()) {
-                            String prefix = getLogPrefix(parent);
-                            if (prefix != null) {
-                                return Optional.of(prefix);
-                            }
-                        }
-                        return Optional.absent();
-                    }
-                });
+            if (branchNameCache == null) {
+                branchNameCache = CacheBuilder.newBuilder().weakKeys().build();
+            }
+            return branchNameCache;
+        }
+    }
+
+    private @CheckForNull String getBranchName(FlowNode node) {
+        Cache<FlowNode, Optional<String>> cache = getBranchNameCache();
+
+        Optional<String> output = cache.getIfPresent(node);
+        if (output != null) {
+            return output.orNull();
+        }
+
+        // We must explicitly check for the current node being the start/end of a parallel branch
+        if (node instanceof BlockEndNode) {
+            output = Optional.fromNullable(getBranchName(((BlockEndNode) node).getStartNode()));
+            cache.put(node, output);
+            return output.orNull();
+        } else if (node instanceof BlockStartNode) { // And of course this node might be the start of a parallel branch
```

Comment:
This trivial conditional check should massively reduce the getPersistentAction calls and improve performance a bit.